File: src/lib/channel-state.ts
import type { ItemListingRow } from "@/hooks/use-item-listings";
import type { ExtensionQueueItem } from "@/hooks/use-extension-queue";
import { safeHref } from "@/lib/safe-url";

// US-3367: one answer to "what is this item doing on this marketplace", for
// the kit tab, the kit status row and the item page card. Two inputs the pages
// already read: the item's listing rows and the seller's extension queue.
// Pure, so the precedence below is tested once rather than guessed at three
// render sites.

export type ChannelState =
  | "live"
  /**
   * Recorded as listed because the form was filled, not because anything saw
   * it go live. The seller opts OUT with "Not listed" (founder, 2026-09-11):
   * a garment recorded as maybe-listed is one they will be told to check when
   * it sells elsewhere; a draft they forgot is one they will not.
   */
  | "unconfirmed"
  | "queued"
  | "delist_queued"
  | "prefilled"
  | "failed"
  | "ended"
  | "sold"
  | "none";

export interface ChannelStatus {
  state: ChannelState;
  row: ItemListingRow | null;
  queueItem: ExtensionQueueItem | null;
  /** The marketplace page, only when the row carries a usable https URL. */
  url: string | null;
  /** ISO time the state started, when known. */
  since: string | null;
}

const PENDING = new Set(["queued", "claimed"]);
const FAILED = new Set(["failed", "expired"]);

function newest<T extends { created_at: string }>(items: T[]): T | null {
  return items.slice().sort((a, b) => b.created_at.localeCompare(a.created_at))[0] ?? null;
}

export function deriveChannelState(
  rows: readonly ItemListingRow[],
  queueItems: readonly ExtensionQueueItem[],
  platform: string,
): ChannelStatus {
  const row = rows.find((r) => r.platform === platform) ?? null;
  const mine = queueItems.filter((it) => it.platform === platform);
  const delist = newest(mine.filter((it) => it.kind === "delist" && PENDING.has(it.status)));
  const list = newest(mine.filter((it) => it.kind === "list"));
  const url = safeHref(row?.listing_url);
  const base: ChannelStatus = {
    state: "none",
    row,
    queueItem: null,
    url,
    since: row?.updated_at ?? null,
  };

  // Precedence, most urgent first. A delist in flight means the garment is
  // gone and the listing may still be live, which beats every other reading.
  if (delist) return { ...base, state: "delist_queued", queueItem: delist, since: delist.created_at };
  if (row?.listing_status === "ended" && row.delist_requested_at) {
    return { ...base, state: "delist_queued", since: row.delist_requested_at };
  }
  if (list && PENDING.has(list.status)) {
    return { ...base, state: "queued", queueItem: list, since: list.created_at };
  }
  if (row?.listing_status === "active") {
    const unconfirmed = Boolean(
      (row.platform_fields as { listed_unconfirmed?: unknown } | null)?.listed_unconfirmed,
    );
    return { ...base, state: unconfirmed && !url ? "unconfirmed" : "live" };
  }
  if (list && FAILED.has(list.status)) {
    return { ...base, state: "failed", queueItem: list, since: list.completed_at ?? list.created_at };
  }
  if (row?.listing_status === "sold") return { ...base, state: "sold" };
  if (row?.listing_status === "ended") return { ...base, state: "ended" };
  if (row?.listing_status === "draft") return { ...base, state: "prefilled" };
  return base;
}

/**
 * The "List on" checklist's defaults. A disabled entry carries the word the
 * label shows in brackets.
 *
 * Prefilled and ended are offered but unchecked: a prefill the seller
 * abandoned may want a second go and an ended listing may want a relist, and
 * both are a decision rather than a default.
 */
export function planListEverywhere(
  platforms: readonly string[],
  statuses: Record<string, ChannelStatus | undefined>,
): { checked: string[]; disabled: Record<string, string> } {
  const checked: string[] = [];
  const disabled: Record<string, string> = {};
  for (const p of platforms) {
    const s = statuses[p]?.state ?? "none";
    if (s === "live") disabled[p] = "live";
    else if (s === "unconfirmed") disabled[p] = "listed?";
    else if (s === "queued") disabled[p] = "queued";
    else if (s === "delist_queued") disabled[p] = "ending";
    else if (s === "none" || s === "failed") checked.push(p);
  }
  return { checked, disabled };
}

File: ios/GradeThread/Marketplaces/ChannelState.swift
import Foundation

/// US-3454: what this item is doing on one marketplace, on the phone.
///
/// ⚠ **MIRROR. The TypeScript is the source.** `deriveChannelState` in
/// `src/lib/channel-state.ts` decides the word the web shows on the item card,
/// the composer's List on panel and the listings table; this is the same
/// precedence for the push sheet and the item screen. `CHANNEL_STATE_PRECEDENCE`
/// there and `precedence` here are compared by
/// `src/test/ios-channel-state-parity.test.ts`, which parses this file as text,
/// so the order below is the contract, not a comment.
///
/// One difference, stated: the phone's `LocalListing` carries no
/// `platform_fields`, so the web's `unconfirmed` (recorded as listed because a
/// form was filled, never seen live) reads as `live` here. The case exists so a
/// future column lands in one place.
enum ChannelState: String, Sendable {
    case live
    case unconfirmed
    case queued
    case delistQueued = "delist_queued"
    case prefilled
    case failed
    case ended
    case sold
    case none

    /// Most urgent first. Mirrors `CHANNEL_STATE_PRECEDENCE` in channel-state.ts.
    static let precedence: [String] = [
        "delist_queued",
        "queued",
        "live",
        "unconfirmed",
        "failed",
        "sold",
        "ended",
        "prefilled",
        "none",
    ]

    /// The word the seller reads. Mirrors the WORDS tables on the web.
    var word: String {
        switch self {
        case .live: return "Live"
        case .unconfirmed: return "Recorded as listed, not confirmed"
        case .queued: return "Queued for your desktop"
        case .delistQueued: return "Ending from your browser"
        case .prefilled: return "Form filled, not confirmed live"
        case .failed: return "Needs you"
        case .ended: return "Ended"
        case .sold: return "Sold here"
        case .none: return ""
        }
    }

    /// A channel the item is already on, or on its way to, cannot be ticked
    /// again: the server would skip it (US-3367) and the count would lie.
    var blocksPush: Bool {
        switch self {
        case .live, .unconfirmed, .queued, .delistQueued: return true
        case .prefilled, .failed, .ended, .sold, .none: return false
        }
    }
}

/// The columns the derivation reads off a listing row. A structural subset so
/// `LocalListing` and a decoded edge row both fit.
struct ChannelRow: Equatable, Sendable {
    let id: String
    let platform: String
    let status: String
    let url: String?
    let delistRequested: Bool

    init(id: String, platform: String, status: String, url: String?, delistRequested: Bool = false) {
        self.id = id
        self.platform = platform
        self.status = status
        self.url = url
        self.delistRequested = delistRequested
    }
}

struct ChannelStatus: Equatable, Sendable {
    let state: ChannelState
    let rowId: String?
    let url: String?
    let queueItemId: String?
    let queueStatus: String?
}

enum ChannelStateDerivation {

    private static let pending: Set<String> = ["queued", "claimed"]
    private static let failed: Set<String> = ["failed", "expired"]

    /// The same precedence as the web, most urgent first: a delist in flight
    /// beats everything (the garment is gone and the listing may still be
    /// live), then a queued list, then the row's own status, then a failed job.
    static func derive(
        rows: [ChannelRow],
        queue: [ExtensionQueueService.QueueItem],
        platform: String
    ) -> ChannelStatus {
        let row = rows.first { $0.platform == platform }
        let mine = queue.filter { $0.platform == platform }
        let delist = newest(mine.filter { $0.kind == "delist" && pending.contains($0.status) })
        let list = newest(mine.filter { $0.kind == "list" })
        let url = row?.url.flatMap { $0.isEmpty ? nil : $0 }

        func status(_ state: ChannelState, queueItem: ExtensionQueueService.QueueItem? = nil) -> ChannelStatus {
            ChannelStatus(
                state: state,
                rowId: row?.id,
                url: url,
                queueItemId: queueItem?.id,
                queueStatus: queueItem?.status
            )
        }

        if let delist { return status(.delistQueued, queueItem: delist) }
        if let row, row.status == "ended", row.delistRequested { return status(.delistQueued) }
        if let list, pending.contains(list.status) { return status(.queued, queueItem: list) }
        if let row, row.status == "active" || row.status == "relisted" { return status(.live) }
        if let list, failed.contains(list.status) { return status(.failed, queueItem: list) }
        if let row, row.status == "sold" { return status(.sold) }
        if let row, row.status == "ended" { return status(.ended) }
        if let row, row.status == "draft" { return status(.prefilled) }
        return status(.none)
    }

    private static func newest(_ items: [ExtensionQueueService.QueueItem]) -> ExtensionQueueService.QueueItem? {
        items.max { $0.createdAt < $1.createdAt }
    }
}
